spec/lib/report_formater/c3_formatter_spec.rb

```diff
@@ -0,0 +1,251 @@
+          ],
+        )
+
+        expect_any_instance_of(described_class).to receive(:build_numeric_chart_simple).once.and_call_original
+        render_report(report)
+        expect(report.chart[:data][:columns][0][1]).to eq(15)
+        expect(report.chart[:data][:columns][0][-1]).to eq(1) if other
+      end
+    end
+
+    it "handles null data in chart column" do
+      report = MiqReport.new(
+        :db          => "Vm",
+        :cols        => %w(name),
+        :include     => {"hardware" => {"columns" => %w(cpu_speed cpu_sockets memory_mb)}},
+        :col_order   => %w(name hardware.cpu_speed hardware.cpu_sockets hardware.memory_mb),
+        :headers     => ["Name", "Hardware CPU Speed", "Hardware Number of CPUs", "Hardware RAM"],
+        :order       => "Ascending",
+        :sortby      => %w(name),
+        :graph       => {:type => "Bar", :mode => "values", :column => "Vm.hardware-cpu_sockets", :count => 10, :other => true},
+        :dims        => 1,
+        :col_options => {},
+        :rpt_options => {},
+        :extras      => {},
+      )
+
+      report.table = Ruport::Data::Table.new(
+        :column_names => %w(name hardware.cpu_speed hardware.cpu_sockets hardware.memory_mb id),
+        :data         => [
+          ["Чук", nil, 4,   6_144, 42],
+          ["Гек", nil, nil, 1_024, 49],
+        ],
+      )
+
+      expect_any_instance_of(described_class).to receive(:build_numeric_chart_simple).once.and_call_original
+      render_report(report)
+    end
+  end
+
+  context '#build_numeric_chart_grouped_2dim' do
+    [true, false].each do |other|
+      it "builds 3d numeric charts #{other ? 'with' : 'without'} 'other'" do
+        report = MiqReport.new(
+          :db          => "Vm",
+          :cols        => %w(os_image_name mem_cpu),
+          :include     => {"ext_management_system" => {"columns" => ["name"]}},
+          :col_order   => %w(ext_management_system.name os_image_name mem_cpu),
+          :headers     => ["Cloud/Infrastructure Provider Name", "OS Name", "Memory"],
+          :order       => "Ascending",
+          :sortby      => %w(ext_management_system.name os_image_name),
+          :group       => "y",
+          :graph       => {:type => "StackedBar", :mode => "values", :column => "Vm-mem_cpu:total", :count => 2, :other => other},
+          :dims        => 2,
+          :col_options => {"name" => {:break_label => "Cloud/Infrastructure Provider : Name: "}, "mem_cpu" => {:grouping => [:total]}},
+          :rpt_options => {:summary => {:hide_detail_rows => false}},
+          :extras      => {},
+        )
+
+        report.table = Ruport::Data::Table.new(
+          :column_names => %w(os_image_name mem_cpu ext_management_system.name id),
+          :data         => [
+            ["linux_centos", 6_144, "MTC-RHEVM-3.0", 67],
+            ["linux_centos", 512,   "MTC-RHEVM-3.0", 167],
+            ["widloze",      1_024, "MTC-RHEVM-3.0", 68],
```

Comment:
Should probably change these.. (and openslack) - funny is OK but this *might* be potentially offensive..